server/rpc.go

```diff
@@ -50,6 +51,85 @@ type Broadcaster interface {
 	Sign([]byte) ([]byte, error)
 }
 
+type BroadcastSessionsManager struct {
+	broadcastSessions []*BroadcastSession
+	broadcastSessMap  map[string]*BroadcastSession
+	sessLock          *sync.Mutex
+}
+
+func (bsm *BroadcastSessionsManager) selectSession() *BroadcastSession {
+	bsm.sessLock.Lock()
+	defer bsm.sessLock.Unlock()
+
+	// minOrchs is maximum number of inflight sessions
+	minOrchs := int(HTTPTimeout / SegLen)
+	if len(bsm.broadcastSessions) < minOrchs {
+		bsm.checkStatus()
+	}
+
+	if len(bsm.broadcastSessions) <= 0 {
+		return nil
+	}
+
+	last := len(bsm.broadcastSessions) - 1
+	sess, sessions := bsm.broadcastSessions[last], bsm.broadcastSessions[:last]
+	bsm.broadcastSessions = sessions
+	return sess
+}
+
+func (bsm *BroadcastSessionsManager) removeSession(session *BroadcastSession) {
+	bsm.sessLock.Lock()
+	delete(bsm.broadcastSessMap, session.OrchestratorInfo.Transcoder)
+	bsm.sessLock.Unlock()
+}
+
+func (bsm *BroadcastSessionsManager) completeSessions(sess *BroadcastSession) {
+	bsm.sessLock.Lock()
+	defer bsm.sessLock.Unlock()
+
+	if _, ok := bsm.broadcastSessMap[sess.OrchestratorInfo.Transcoder]; ok {
+		bsm.broadcastSessions = append(bsm.broadcastSessions, sess)
+	}
+}
+
+// func (bsm *BroadcastSessionsManager) checkStatus(node *core.LivepeerNode, pl core.PlaylistManager) {
+func (bsm *BroadcastSessionsManager) checkStatus() {
+	// newBroadcastSessions, err := selectOrchs(node, pl)
+	// if err != nil {
+	// 	return
+	// }
+
+	newBroadcastSessions := []*BroadcastSession{}
+
+	// include all O's being used successfully in refreshed list
+	if len(newBroadcastSessions) > 0 {
```

Comment:
I think the flow for the rest of this function can be simplified to:

```golang
uniqueSessions := make([]*BroadcastSession, len(newBroadcastSessions)) // minimize allocations
for _, sess := range newBroadcastSessions {
  if _, ok := bsm.broadcastSessMap[sess.OrchestratorInfo.Transcoder]; ok {
    continue
  }
  uniqueSessions = append(uniqueSessions, sess)
  bsm.broadcastSessMap[sess.OrchestratorInfo.Transcoder] = sess
}
bsm.broadcastSessions = append(uniqueSessions, bsm.broadcastSessions...)
```

Reply:
Much more concise! This works